wazo_auth/services/saml.py

```diff
@@ -117,57 +123,148 @@ def prepare_redirect_response(
     ):
         saml_session_id = secrets.token_urlsafe(16)
         client = self.get_client(domain)
-        reqid, info = client.prepare_for_authenticate(relay_state=domain)
 
-        self._outstanding_requests[reqid] = SamlAuthContext(
+        relay_state: str = base64.urlsafe_b64encode(
+            hashlib.sha256(saml_session_id.encode()).digest()
+        ).decode()
+        req_id, info = client.prepare_for_authenticate(relay_state=relay_state)
+
+        self._outstanding_requests[req_id] = SamlAuthContext(
             saml_session_id,
             redirect_url,
             domain,
+            relay_state,
         )
         location = [i for i in info['headers'] if i[0] == 'Location'][0][1]
         return location, saml_session_id
 
+    def _decode_saml_response(
+        self, saml_client: Saml2Client, saml_response: str, conv_info: dict[str, Any]
+    ) -> None | AuthnResponse:
+        return saml_client.parse_authn_request_response(
+            saml_response,
+            BINDING_HTTP_POST,
+            self._outstanding_requests,
+            None,
+            conv_info=conv_info,
+        )
+
+    def _find_session_by_relay_state(
+        self, relay_state: str
+    ) -> tuple[SamlAuthContext, SamlAuthContext] | tuple[None, None]:
```

Comment:
```suggestion
    ) -> tuple[SamlAuthContext, Requestid] | tuple[None, None]:
```
error in type annotation